File: src/data/Trees.cpp
#include <iostream>

#include "data/Trees.h"

/**
 * @brief X-Tree Construction given precomputed components.
 * 
 * @param ranker Prebuilt ranker table
 * @param shortlex Precomputed `ShortlexResult` of a pattern string
 * @param text Text.
 * @return `std::shared_ptr<Node>` Root of the tree.
 */
std::shared_ptr<Node> Trees::buildXTree(const RankerTable& ranker, const ShortlexResult& shortlex, const std::string& text) {
    std::shared_ptr<Node> root = std::make_shared<Node>(RankerTable::INF, RankerTable::INF, 0);
    std::unordered_map<int, std::shared_ptr<Node>> nodes;
    nodes[RankerTable::INF] = root;
    
    std::cout << "Building X-tree...\n";

    // Copy out s_p
    std::deque<std::set<char>> s_p;
    std::set<char> s;
    for(int i = 0; i < shortlex.stackForm.size(); i++){
        s = shortlex.stackForm.at(i);
        s_p.push_back(s);
    }

    // ln 4-6
    int last_arch = 0;
    for(int i = 0; i < shortlex.arch_ends.size(); i++){
        for(int j = 0; j < shortlex.arch_ends.at(i) - last_arch; j++){
            s_p.back().erase(shortlex.shortlexNormalForm.at(j+last_arch));
            if(s_p.back().empty()){ s_p.pop_back(); }
        }
        last_arch = shortlex.arch_ends.at(i);
    }

    // ln 7-21
    int parent; int x_rank; int min_x_rank;
    std::deque<std::set<char>> sp_p;
    std::unordered_map<int, int> r; std::set<char> S;
    char sigma;
    for(int i = 0; i < static_cast<int>(text.size()); i++){
        parent = -1;
        for(char a : shortlex.alphabet){
            x_rank = ranker.getX(i, a);
            if(x_rank > parent) parent = x_rank;
        }

        // ln 9-18
        if(nodes.count(parent) == 0) {
            nodes[parent] = std::make_shared<Node>(parent, parent, 0);
            std::cout << "Generate new node " << parent << std::endl;

            // line 11: s'_p <- copy(s_p)
            for(auto entry = s_p.begin(); entry != s_p.end(); entry++){
                sp_p.push_back(*entry);
            }

            // line 12: T_X(T).r(i) <- i
            r[i]= i;

            // line 13: while s'_p is not empty
            while(!sp_p.empty()){
                // line 14: S <- peek(s'_p)
                S = sp_p.back();

                // line 15: sigma = arg min (R_X(T, r(i), c))
                min_x_rank = -1;
                for(char c : S) {
                    x_rank = ranker.getX(r[i], c);
                    if (min_x_rank == -1 || x_rank < min_x_rank) { 
                        min_x_rank = x_rank;
                        sigma = c;
                    }
                }

                // line 16: T_X(T).r(i) <- R_X(T, T_X(T).r(i), sigma)
                r[i] = ranker.getX(r[i], sigma);
                if (r[i] == RankerTable::INF) {
                    sp_p.clear();
                    break;
                }

                // line 17: pop sigma from s'_p
                sp_p.back().erase(sigma);
                if (sp_p.back().empty()) {
                    sp_p.pop_back();
                }
            }

            // line 18: implemented within line 19
        }

        // line 19: extend end point of T_X(`T`).chld(parent) by one
        if(nodes.count(i) != 0){
            nodes[i]->parent = nodes[parent];
            nodes[parent]->children.push_back(nodes[i]);    // line 18: T_X(T).chld(parent) <- [i, i)
            nodes[i]->r = r[i];
            std::cout << "Set parent of " << i << " to " << parent << std::endl;
        }
    }

    // Clean up
    for(auto pair : nodes){
        int index = pair.first;
        std::shared_ptr<Node> node = pair.second;
        if(node->parent == nullptr && node != root){
            node->parent = root;
            root->children.push_back(node);
            node->r = r[index];
            std::cout << "Set parent of " << pair.first << " to " << RankerTable::INF << std::endl;
        }
    }

    std::cout << "End of X-tree construction\n\n"; // Double lb is intended
    return root;
}

/**
 * @brief Y-Tree Construction given precomputed components.
 * 
 * @param ranker Prebuilt ranker table
 * @param shortlex Precomputed `ShortlexResult` of a pattern string
 * @param text Text.
 * @return `std::shared_ptr<Node>` Root of the tree.
 */
std::shared_ptr<Node> Trees::buildYTree(const RankerTable& ranker, const ShortlexResult& shortlex, const std::string& text) {
    std::shared_ptr<Node> root = std::make_shared<Node>(-1, -1, 0);
    std::unordered_map<int, std::shared_ptr<Node>> nodes;
    nodes[-1] = root;
    
    std::cout << "Building Y-tree...\n";

    // Copy out s_p (in reverse order) << not sure if reversing is mandatory
    std::vector<std::set<char>> s_p; std::set<char> s;
    for(int i = 0; i < shortlex.stackForm.size(); i++){
        s = shortlex.stackForm.at(shortlex.stackForm.size() - i - 1);
        s_p.push_back(s);
    }

    // ln 4-6 (Slight rework as well)
    int deleted_chars = 0;
    for(int i = 0; i < shortlex.universality; i++){
        while(deleted_chars < shortlex.alphabet.size()){
            deleted_chars += s_p.back().size();
            s_p.pop_back();
        }
        deleted_chars = 0;
    }

    // ln 7-21
    int parent; int y_rank; int max_y_rank;
    std::vector<std::set<char>> sp_p;
    std::unordered_map<int, int> r; std::set<char> S;
    unsigned char sigma;
    for(int i = static_cast<int>(text.size()); i > 0; i--){
        parent = RankerTable::INF;
        for(auto a : shortlex.alphabet){
            y_rank = ranker.getY(i, a);
            if(y_rank < parent) parent = y_rank;
        }

        if(parent != -1) parent--;

        // ln 9-18
        if(nodes.count(parent) == 0){
            nodes[parent] = std::make_shared<Node>(parent, parent, 0);
            std::cout << "Generate new node " << parent << std::endl;

            // line 11: s'_p <- copy(s_p)
            for(auto entry = s_p.begin(); entry != s_p.end(); entry++){
                sp_p.push_back(*entry);
            }

            // line 12: T_Y(T).r(i) <- i
            r[i] = i;

            // line 13: while s'_p is not empty
            while(!sp_p.empty()){
                // line 14: S <- peek(s'_p)
                S = sp_p.back();

                // line 15: sigma = arg min (R_Y(T, r(i), c))
                max_y_rank = RankerTable::INF;
                for(char c : S) {
                    y_rank = ranker.getY(r[i], c);
                    if (max_y_rank == RankerTable::INF || y_rank > max_y_rank) {
                        max_y_rank = y_rank;
                        sigma = c;
                    }
                }

                // line 16: T_Y(T).r(i) <- R_Y(T, T_Y(T).r(i), sigma)
                r[i] = ranker.getY(r[i], sigma);
                if (r[i] < 0) {
                    sp_p.clear();
                    break;
                }

                // line 17: pop sigma from s'_p
                sp_p.back().erase(sigma);
                if(sp_p.back().empty()) {
                    sp_p.pop_back();
                }
            }

            // line 18: implemented within line 19
        }

        // line 19: extend end point of T_Y(T).chld(parent) by one
        if(nodes.count(i) != 0){
            nodes[i]->parent = nodes[parent];
            nodes[parent]->children.push_back(nodes[i]);    // line 18: T_Y(T).chld(parent) <- [i, i)
            nodes[i]->r = r[i];
            std::cout << "Set parent of " << i << " to " << parent << std::endl;
        }
    }

    // Clean up
    for(auto pair : nodes){
        int index = pair.first;
        std::shared_ptr<Node> node = pair.second;
        if(node->parent == nullptr && node != root){
            node->parent = root;
            root->children.push_back(node);
            node->r = r[index];
            std::cout << "Set parent of " << pair.first << " to -1" << std::endl;
        }
    }

    std::cout << "End of Y-tree construction\n\n"; // Double lb is intended
    return root;
}

File: animations/data/Trees.py
from collections import deque
from typing import Dict, Set
from data.Node import Node


class Trees:
    @staticmethod
    def build_x_tree(ranker, shortlex, text: str) -> Node:
        print("Building X-tree...")
        INF = float('inf')
        root = Node(INF, INF, 0)
        nodes: Dict[int, Node] = {INF: root}

        # Copy s_p from shortlex.stackForm
        s_p = deque(set(block) for block in shortlex.stackForm)

        # ln 4-6: consume blocks based on arch ends
        last_arch = 0
        for end in shortlex.arch_ends:
            for j in range(end - last_arch):
                ch = shortlex.shortlexNormalForm[j + last_arch]
                s_p[-1].discard(ch)
                if not s_p[-1]:
                    s_p.pop()
            last_arch = end

        # ln 7-21: main construction
        r: Dict[int, int] = {}
        for i in range(len(text)):
            parent = -69
            for a in shortlex.alphabet:
                x_rnk = ranker.get_x(i, a)
                if x_rnk > parent:
                    parent = x_rnk

            if parent not in nodes:
                nodes[parent] = Node(parent, parent, 0)
                print(f"Generate new node {parent}")

                # Copy s_p to sp_p
                sp_p = deque(set(block) for block in s_p)
                r[i] = i

                while sp_p:
                    S = sp_p[-1]
                    min_x_rnk = None
                    sig = None
                    for c in S:
                        x_rnk = ranker.get_x(r[i], c)
                        if min_x_rnk is None or x_rnk < min_x_rnk:
                            min_x_rnk = x_rnk
                            sig = c
                    r[i] = ranker.get_x(r[i], sig)
                    S.discard(sig)
                    if not S:
                        sp_p.pop()

            if i in nodes:
                nodes[i].parent = nodes[parent]
                nodes[parent].children.append(nodes[i])
                print(f"Set parent of {i} to {parent}")

        # Clean up
        for i, node in nodes.items():
            if node.parent is None and node is not root:
                node.parent = root
                root.children.append(node)
                print(f"Set parent of {i} to root")

        return root

    @staticmethod
    def build_y_tree(ranker, shortlex, text: str) -> Node:
        print("Building Y-tree...")
        root = Node(-1, -1, 0)
        nodes: Dict[int, Node] = {-1: root}

        # Copy out s_p (in reverse order)
        s_p = [set(shortlex.stackForm[len(shortlex.stackForm) - i - 1]) for i in range(len(shortlex.stackForm))]

        # Remove universality blocks
        deleted_chars = 0
        for _ in range(shortlex.universality):
            while deleted_chars < len(shortlex.alphabet):
                deleted_chars += len(s_p[-1])
                s_p.pop()
            deleted_chars = 0

        r: Dict[int, int] = {}
        for i in reversed(range(1, len(text) + 1)):
            parent = float('inf')
            for a in shortlex.alphabet:
                y_rnk = ranker.get_y(i, a)
                if y_rnk < parent:
                    parent = y_rnk
            if parent != -1:
                parent -= 1

            if parent not in nodes:
                nodes[parent] = Node(parent, parent, 0)
                print(f"Generate new node {parent}")

                sp_p = [set(s) for s in s_p]
                r[i] = i

                while sp_p:
                    S = sp_p[-1]
                    max_y_rnk = -1
                    sig = None
                    for c in S:
                        y_rnk = ranker.get_y(r[i], c)
                        if max_y_rnk == -1 or y_rnk > max_y_rnk:
                            max_y_rnk = y_rnk
                            sig = c
                    r[i] = ranker.get_y(r[i], sig)
                    S.remove(sig)
                    if not S:
                        sp_p.pop()

            if i in nodes:
                nodes[i].parent = nodes[parent]
                nodes[parent].children.append(nodes[i])
                print(f"Set parent of {i} to {parent}")

        for i, node in nodes.items():
            if node.parent is None and node is not root:
                node.parent = root
                root.children.append(node)
                print(f"Set parent of {i} to root")

        return root